lighthouse-core/audits/mainthread-work-breakdown.js

```diff
@@ -0,0 +1,172 @@
+  'Other': group.other,
+  'Rasterize Paint': group.painting,
+  'Recalculate Style': group.styleLayout,
+  'Request Animation Frame': group.scripting,
+  'Request Idle Callback': group.scripting,
+  'Request Main Thread Frame': group.painting,
+  'Image Resize': group.images,
+  'Finish Loading': group.loading,
+  'Receive Data': group.loading,
+  'Receive Response': group.loading,
+  'Send Request': group.loading,
+  'Run Microtasks': group.scripting,
+  'Schedule Style Recalculation': group.styleLayout,
+  'Scroll': group.compositing,
+  'Task': group.other,
+  'Timer Fired': group.scripting,
+  'Install Timer': group.scripting,
+  'Remove Timer': group.scripting,
+  'Timestamp': group.scripting,
+  'Update Layer': group.compositing,
+  'Update Layer Tree': group.compositing,
+  'User Timing': group.scripting,
+  'Create WebSocket': group.scripting,
+  'Destroy WebSocket': group.scripting,
+  'Receive WebSocket Handshake': group.scripting,
+  'Send WebSocket Handshake': group.scripting,
+  'XHR Load': group.scripting,
+  'XHR Ready State Change': group.scripting,
+};
+
+class PageExecutionTimings extends Audit {
+  /**
+   * @return {!AuditMeta}
+   */
+  static get meta() {
+    return {
+      category: 'Performance',
+      name: 'mainthread-work-breakdown',
+      description: 'Main thread work breakdown',
+      informative: true,
+      helpText: 'Consider reducing the time spent parsing, compiling and executing JS.' +
+        'You may find delivering smaller JS payloads helps with this.',
+      requiredArtifacts: ['traces'],
+    };
+  }
+
+  /**
+   * @param {!Array<TraceEvent>} trace
+   * @return {!Map<string, number>}
+   */
+  static getExecutionTimingsByCategory(trace) {
+    const timelineModel = new DevtoolsTimelineModel(trace);
+    const bottomUpByName = timelineModel.bottomUpGroupBy('EventName');
+
+    const result = new Map();
+    bottomUpByName.children.forEach((event, eventName) =>
+      result.set(eventName, event.selfTime));
+
+    return result;
+  }
+
+  /**
+   * @param {!Artifacts} artifacts
+   * @return {!AuditResult}
+   */
+  static audit(artifacts) {
+    const trace = artifacts.traces[PageExecutionTimings.DEFAULT_PASS];
+    const executionTimings = PageExecutionTimings.getExecutionTimingsByCategory(trace);
+    let totalExecutionTime = 0;
+
+    const extendedInfo = {};
+    const results = Array.from(executionTimings).map(([eventName, duration]) => {
+      totalExecutionTime += duration;
+      extendedInfo[eventName] = duration;
+
+      return {
+        category: eventName,
+        group: taskToGroup[eventName],
+        duration: Util.formatMilliseconds(duration, 1),
+      };
+    });
+
+    const headings = [
+      {key: 'category', itemType: 'text', text: 'Category'},
+      {key: 'group', itemType: 'text', text: 'Task Category'},
+      {key: 'duration', itemType: 'text', text: 'Time spent'},
+    ];
+    const tableDetails = PageExecutionTimings.makeTableDetails(headings, results);
```

Comment:
can you sort this data?

the bottomupbyname was already sorted by Duration so we dont need to do that, but let's sort by each `group`'s total time spent.

so basically we'll have all the Script Evaluation work (as that will likely be the top one), with its tasks then sorted in decreasing order of time, then the rest of the groups.

Reply:
before and after: 
![image](https://user-images.githubusercontent.com/39191/32254711-cfd7680c-be5f-11e7-9396-a68f24db9abe.png)

(different data)

Reply:
here's the patch for all this work:

```diff
diff --git a/lighthouse-core/audits/mainthread-work-breakdown.js b/lighthouse-core/audits/mainthread-work-breakdown.js
index 0dd563e4..c3fd7461 100644
--- a/lighthouse-core/audits/mainthread-work-breakdown.js
+++ b/lighthouse-core/audits/mainthread-work-breakdown.js
@@ -139,22 +139,31 @@ class PageExecutionTimings extends Audit {
     let totalExecutionTime = 0;
 
     const extendedInfo = {};
+    const categoryTotals = {};
+
     const results = Array.from(executionTimings).map(([eventName, duration]) => {
       totalExecutionTime += duration;
       extendedInfo[eventName] = duration;
+      const groupName = taskToGroup[eventName];
+
+      const catTotal = categoryTotals[groupName] || 0;
+      categoryTotals[groupName] = catTotal + duration;
 
       return {
         category: eventName,
-        group: taskToGroup[eventName],
+        group: groupName,
         duration: Util.formatMilliseconds(duration, 1),
       };
     });
 
     const headings = [
-      {key: 'category', itemType: 'text', text: 'Category'},
-      {key: 'group', itemType: 'text', text: 'Task Category'},
+      {key: 'group', itemType: 'text', text: 'Category'},
+      {key: 'category', itemType: 'text', text: 'Work'},
       {key: 'duration', itemType: 'text', text: 'Time spent'},
     ];
+
+    results.stableSort((a, b) => categoryTotals[b.group] - categoryTotals[a.group]);
     const tableDetails = PageExecutionTimings.makeTableDetails(headings, results);
 
     return {

```

i swear at some point ill stop editing your PRs but i enjoy them so much i want to try some things out! :)  sorry!

Reply:
no problem! Makes my job much easier anyway so keep editing away! 😄 💯